### DISCUSSION TITLE
Allowing CPM to (further) facilitate vendoring of libraries

### DISCUSSION THREAD

**Discussion by studoot** [cpm-cmake/CPM.cmake#270] [Ideas]
As the title says, I'm looking to use CPM to vender 3rd party libraries. Placing `CPM_SOURCE_CACHE` within my application's repository gets me 95% of the way there, but there's one little speed hump that I'd like to eliminate if possible.

Currently, when CPM clones a Git repository into `CPM_SOURCE_CACHE`, the repository (i.e. the `.git` directory) is retained - that's just what the underlying CMake functions do.  This is an issue when adding the library to my application's repository, as shown by the error message when `git add`ing it:

```
hint: You've added another git repository inside your current repository.
hint: Clones of the outer repository will not contain the contents of
hint: the embedded repository and will not know how to obtain it.
hint: If you meant to add a submodule, use:
hint:
hint:   git submodule add <url> externals/cpm/catch2/5b0288370c9df775ca7b6073320e23697b91be84
hint:
hint: If you added this path by mistake, you can remove it from the
hint: index with:
hint:
hint:   git rm --cached externals/cpm/catch2/5b0288370c9df775ca7b6073320e23697b91be84
hint:
hint: See "git help submodule" for more information.
```

Obviously, I can manually remove the `.git` directory, but I've been thinking about how achieving this might be incorporated into CPM. I've thought of a few mechanisms that could be used to achieve this:

1. _Remove the `.git` directory after `FetchContent_Populate` has been used to clone the repository_

    This is just what I do manually and is pretty simple to accomplish - I've got a [PoC of this approach](https://github.com/studoot/CPM.cmake/blob/6f7f06221c667feced49bc9a80e44e401d7321da/cmake/CPM.cmake#L597-L607), which removes git repo directories when a variable (`CPM_GIT_REMOVE`) is set. It just feels... inelegant.

2. _Use `git subtree` or `submodule` to incorporate the library's repository into my application's repository._

    This has the advantage of retaining where the library came from. However, it modifies the application's repository by running CMake, which doesn't feel like the right thing to do.

3. _Clone the repository to a temp. directory and use `git archive` to copy out the desired commit's files_

    The canonical way to do this would be to do something like:

    ```
    git -C repo-dir archive --format=tar some-tag | tar -x -C tree-dir
    ```

    However, given Windows, you'd probably want to archive to a file and then use CMake to extract the archive to the desired directory. This is starting to feel a bit like too many moving parts to me...

4. _#3, but use `git checkout-index` instead of `git archive`_

    Forget about #3 - `checkout-index` does what's needed directly.

I guess the question is - is this something that could go in CPM? I'm happy enough to come up with a PR to implement, but it's more about whether this is something you'd be happy having in CPM.

**Comment by TheLartians** [comment]
If I understand the use-case correctly, you essentially want to create a directory containing all dependencies without their according version control data and commit that into a monorepo?

In general I'm not sure if this fits into the scope of CPM.cmake, especially as we are generally interested in keeping the version control information, for example to implement #274. Adding such options would increase the complexity of similar additional features.

Imo I think a better approach would be to write a short script to remove any version control directories inside the local `CPM_SOURCE_CACHE` that runs after the main CMake routines are invoked.

**Reply by studoot** [reply]
Given <blockquote>we are generally interested in keeping the version control information</blockquote>, I'd have to concur.
